core/src/main/scala/org/apache/spark/BarrierCoordinator.scala

```diff
@@ -0,0 +1,199 @@
+    val barrierEpoch = barrierEpochByStageIdAndAttempt.putIfAbsent((stageId, stageAttemptId),
+      new AtomicInteger(0))
+    if (barrierEpoch == null) {
+      barrierEpochByStageIdAndAttempt.get((stageId, stageAttemptId))
+    } else {
+      barrierEpoch
+    }
+  }
+
+  /**
+   * Send failure to all the blocking barrier sync requests from a stage attempt with proper
+   * failure message.
+   */
+  private def failAllSyncRequests(
+      syncRequests: ArrayBuffer[RpcCallContext],
+      message: String): Unit = {
+    syncRequests.foreach(_.sendFailure(new SparkException(message)))
+  }
+
+  /**
+   * Finish all the blocking barrier sync requests from a stage attempt successfully if we
+   * have received all the sync requests.
+   */
+  private def maybeFinishAllSyncRequests(
+      syncRequests: ArrayBuffer[RpcCallContext],
+      numTasks: Int): Boolean = {
+    if (syncRequests.size == numTasks) {
+      syncRequests.foreach(_.reply(()))
+      return true
+    }
+
+    false
+  }
+
+
+  override def receiveAndReply(context: RpcCallContext): PartialFunction[Any, Unit] = {
+    case RequestToSync(numTasks, stageId, stageAttemptId, taskAttemptId, barrierEpoch) =>
+      // Check the barrier epoch, to see which barrier() call we are processing.
+      val currentBarrierEpoch = getOrInitBarrierEpoch(stageId, stageAttemptId)
+      logInfo(s"Current barrier epoch for Stage $stageId (Attempt $stageAttemptId) is" +
+        s"$currentBarrierEpoch.")
+      if (currentBarrierEpoch.get() != barrierEpoch) {
+        context.sendFailure(new SparkException(s"The request to sync of Stage $stageId (Attempt " +
+          s"$stageAttemptId) with barrier epoch $barrierEpoch has already finished. Maybe task " +
+          s"$taskAttemptId is not properly killed."))
+      } else {
+        val syncRequests = getOrInitSyncRequests(stageId, stageAttemptId)
+        // If this is the first sync message received for a barrier() call, init a timer to ensure
+        // we may timeout for the sync.
+        if (syncRequests.isEmpty) {
+          timer.schedule(new TimerTask {
+            override def run(): Unit = {
+              // Timeout for current barrier() call, fail all the sync requests.
+              val requests = getOrInitSyncRequests(stageId, stageAttemptId)
+              failAllSyncRequests(requests, "The coordinator didn't get all barrier sync " +
+                s"requests for barrier epoch $barrierEpoch from Stage $stageId (Attempt " +
+                s"$stageAttemptId) within ${timeout}s.")
+              cleanupSyncRequests(stageId, stageAttemptId)
+              // The global sync fails so the stage is expected to retry another attempt, all sync
+              // messages come from current stage attempt shall fail.
+              currentBarrierEpoch.set(-1)
+            }
+          }, timeout * 1000)
+        }
+
+        syncRequests += context
```

Comment:
although very unlikely, shall we add an assert that `syncRequests.length == numTasks`? Just in case we have a bug and some barrier tasks have a different value of `numTasks`.

Reply:
We don't remember the `numTasks` in BarrierCoordinator, if it really worth that then we have to use another map to store the information.

Reply:
each barrier task remembers numTask, so here we can make sure  barrier tasks of same group would have same `numTasks`